### PR TITLE
User group API and cache.

### PR DESCRIPTION
**Pull request by mdwn**
The user group has been added to the API and to the cache.

### CODE REVIEW

**Comment by tobiaszheller** [comment]
@mdwn do you have RFD to that PR? I was searching to better understand context but couldn't find one. 
I think for future it would be good to divide that kind of PR into smaller PRs, one for proto, one for each method etc.

**Comment by mdwn** [comment]
https://github.com/gravitational/teleport.e/pull/609 Here's the RFD. This gets difficult to split up too much because frankly it's almost entirely boilerplate. This was split from another PR as is, and it needs to be backported to v12 -- we could be looking at dozens of PRs for what is effectively adding in CRUD operations for a new object.

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -5510,6 +5510,60 @@ func (a *ServerWithRoles) DeleteAllSAMLIdPServiceProviders(ctx context.Context)
...
+func (a *ServerWithRoles) GetUserGroup(ctx context.Context, name string) (types.UserGroup, error) {
+	if err := a.action(apidefaults.Namespace, types.KindUserGroup, types.VerbRead); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return a.authServer.GetUserGroup(ctx, name)
+}
+
+// CreateUserGroup creates a new user group resource.
+func (a *ServerWithRoles) CreateUserGroup(ctx context.Context, g types.UserGroup) error {
```
Comment (tobiaszheller) [inline lib/auth/auth_with_roles.go:5532]: is validation of inputs is done on other layer?

**Review by tobiaszheller — APPROVED** [review]
lgtm, left 1 question

File: lib/auth/auth_with_roles.go (hunk 1 above)
Comment (mdwn) [reply lib/auth/auth_with_roles.go:5532]: Yes, that's correct.

File: lib/cache/cache.go [hunk 2]
```diff
@@ -152,6 +153,7 @@ func ForProxy(cfg Config) Config {
 		{Kind: types.KindInstaller},
 		{Kind: types.KindKubernetesCluster},
 		{Kind: types.KindSAMLIdPServiceProvider},
+		{Kind: types.KindUserGroup},
```
Comment (fspmarshall) [inline lib/cache/cache.go:156]: Keep in mind that new resources can only be added to non-auth caches during major versions, meaning this cannot be backported to any existing release.

We're working on ways to change this, but for now, adding a resource in a peripheral cache during a minor/patch update breaks our compatibility guarantees.

**Review by fspmarshall — APPROVED** [review]

**Pull request merged**